apps/enterprise-knowledge-retrieval/enterprise_knowledge_retrieval.ipynb

```diff
@@ -0,0 +1,1750 @@
+    "Evaluation can only be \"Correct\", \"Incorrect\", and \"Unable to evaluate\"\n",
+    "\n",
+    "Example 1:\n",
+    "\n",
+    "Question: What is the cost cap for the 2023 season of Formula 1?\n",
+    "\n",
+    "Answer: The cost cap for 2023 is 95m USD.\n",
+    "\n",
+    "Evaluation: Correct\n",
+    "\n",
+    "Example 2:\n",
+    "\n",
+    "Question: What is Thomas Dolby known for?\n",
+    "\n",
+    "Answer: Inventing electricity\n",
+    "\n",
+    "Evaluation: Incorrect\n",
+    "\n",
+    "Begin!'''\n",
+    "\n",
+    "gpt_evaluator_message = '''\n",
+    "Question: QUESTION_HERE\n",
+    "\n",
+    "Answer: ANSWER_HERE\n",
+    "\n",
+    "Evaluation:'''"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 94,
+   "id": "8a351793",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "evaluation_output = []"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 95,
+   "id": "3b286e5f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for pair in question_answer_pairs:\n",
+    "    \n",
+    "    message = gpt_evaluator_message.replace('QUESTION_HERE',pair[0]).replace('ANSWER_HERE',pair[1])\n",
+    "    evaluation = openai.ChatCompletion.create(model='gpt-3.5-turbo',messages=[{\"role\":\"system\",\"content\":gpt_evaluator_system},{\"role\":\"user\",\"content\":message}],temperature=0)\n",
+    "    #print(evaluation)\n",
+    "    \n",
+    "    evaluation_output.append((pair[0],pair[1],evaluation['choices'][0]['message']['content']))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 97,
+   "id": "f18678e0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def collate_results(x):\n",
+    "    text = x.lower()\n",
+    "    \n",
+    "    if 'incorrect' in text:\n",
+    "        return 'incorrect'\n",
+    "    elif 'correct' in text:\n",
+    "        return 'correct'\n",
+    "    else: \n",
+    "        return 'unable to answer'"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 98,
+   "id": "cd29fc90",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "correct             50\n",
+       "unable to answer    28\n",
+       "incorrect           5 \n",
+       "Name: evaluation, dtype: int64"
+      ]
+     },
+     "execution_count": 98,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "eval_df = pd.DataFrame(evaluation_output)\n",
+    "eval_df.columns = ['question','answer','evaluation']\n",
+    "# Replacing all the \"unable to evaluates\" with \"unable to answer\"\n",
+    "eval_df['evaluation'] = eval_df['evaluation'].apply(lambda x: collate_results(x))\n",
+    "eval_df.evaluation.value_counts()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "566053d6",
+   "metadata": {},
+   "source": [
+    "#### Analysis\n",
+    "\n",
+    "We ended up with a 60% first hit rate, which isn't great, but we've got a baseline to work from.\n",
```

Comment:
I wasn't quite able to make it to the end, but did these numbers look right? Is it able to grade accurately?

Reply:
Yes, i've added a "Knowledge" tool so it can use its own knowledge, so this works end to end now. If you rerun it shouldn't take long to go end-to-end.